### PR TITLE
chore: refactor shell to individual files

### PR DESCRIPTION
**Pull request by sean-brydon**
## What does this PR do?

This PR is one of many that refactors the Shell component out of one file and into many sub components. This PR is just extracting with no performance improvements. They will happen in a follow up PR 

## Mandatory Tasks (DO NOT REMOVE)

- [ ] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

### CODE REVIEW

File: packages/features/auth/lib/hooks/useRedirectToLoginIfUnauthenticated.tsx
Comment (sean-brydon) [inline packages/features/auth/lib/hooks/useRedirectToLoginIfUnauthenticated.tsx:1]: Move to a hook that can live outside of shell

File: packages/features/auth/lib/hooks/useRedirectToOnboardingIfNeeded.tsx
Comment (sean-brydon) [inline packages/features/auth/lib/hooks/useRedirectToOnboardingIfNeeded.tsx:1]: Same goes as above with its helper fuctions

File: packages/features/ee/support/lib/intercom/useIntercom.ts
Comment (sean-brydon) [inline packages/features/ee/support/lib/intercom/useIntercom.ts:1]: Create a hook that boots intercom - no need for this to live directly in shell either. Means we can control when the "me" query is called much easier i the future as it was being called on every render

File: packages/features/shell/Shell.tsx
Comment (sean-brydon) [inline packages/features/shell/Shell.tsx:1]: Wow thats a lot less lines

File: packages/features/shell/banners/LayoutBanner.tsx [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+export const BannerContainer: React.FC<BannerContainerProps> = ({ banners }) => {
+  return (
+    <div className="sticky top-0 z-10 w-full divide-y divide-black">
+      {Object.entries(banners).map(([key, data]) => {
+        const BannerComponentToRender = BannerComponent[key as BannerType];
+        if (BannerComponentToRender) {
+          return <BannerComponentToRender key={key} data={data} />;
+        }
```
Comment (sean-brydon) [inline packages/features/shell/banners/LayoutBanner.tsx:64]: The old confused me - its exactly the same as this but way more simple
```ts
           {Object.keys(banners).map((key) => {
              if (key === "teamUpgradeBanner") {
                const Banner = BannerComponent[key];
                return <Banner data={banners[key]} key={key} />;
              } else if (key === "orgUpgradeBanner") {
                const Banner = BannerComponent[key];
                return <Banner data={banners[key]} key={key} />;
              } else if (key === "verifyEmailBanner") {
                const Banner = BannerComponent[key];
                return <Banner data={banners[key]} key={key} />;
              } else if (key === "adminPasswordBanner") {
                const Banner = BannerComponent[key];
                return <Banner data={banners[key]} key={key} />;
              } else if (key === "impersonationBanner") {
                const Banner = BannerComponent[key];
                return <Banner data={banners[key]} key={key} />;
              } else if (key === "calendarCredentialBanner") {
                const Banner = BannerComponent[key];
                return <Banner data={banners[key]} key={key} />;
              } else if (key === "invalidAppCredentialBanners") {
                const Banner = BannerComponent[key];
                return <Banner data={banners[key]} key={key} />;
              }
            })}
```

File: packages/features/shell/banners/useBanners.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      Object.entries(banners).filter(([_, value]) => {
+        return value && (!Array.isArray(value) || value.length > 0);
+      });
+    return (activeBanners?.length ?? 0) * TOP_BANNER_HEIGHT;
+  }, [banners]);
+
+  return bannersHeight;
+};
+
+const useBanners = () => {
```
Comment (sean-brydon) [inline packages/features/shell/banners/useBanners.ts:41]: Lets just have one hook to get all banners and calculate the height

**Review by joeauyeung — CHANGES_REQUESTED** [review]
Great refactor @sean-brydon

## Test Cases
- Desktop view ✅ 
- Reduced side bar ✅ 
- Mobile nav ❌ 
- Unauthed redirect to login page ✅ 
- Onboarding trigger ✅ 

## Blocking
- If we're refactoring right now can we abstract some of the hooks logic so we can prepare for the app router migration? If it's too big then I'm ok having this in a follow up PR. FYI @hbjORbj 
- In the mobile nav container when clicking `More` I get this error
CleanShot 2024-10-01 at 10 27 25@2x


## Non-blocking
- Have you considered using `useCallback` at all to help with preventing rerenders? I'm also ok with having it in a follow up PR

File: packages/features/shell/user-dropdown/ProfileDropdown.tsx [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        </button>
+      </DropdownMenuTrigger>
+
+      <DropdownMenuPortal>
+        <DropdownMenuContent
+          align="start"
+          onInteractOutside={() => {
+            setMenuOpen(false);
+          }}
+          className="min-w-56 hariom group overflow-hidden rounded-md">
```
[the remark this answers is not in the document]
Comment (sean-brydon) [reply packages/features/shell/user-dropdown/ProfileDropdown.tsx:72]: Thats a great question lol

**Comment by sean-brydon** [comment]
> * Have you considered using `useCallback` at all to help with preventing rerenders? I'm also ok with having it in a follow up PR

100%! As i mentioned in the PR description - i've not made any perf improvements here. Just making the component more manageable! 

> If we're refactoring right now can we abstract some of the hooks logic so we can prepare for the app router migration? If it's too big then I'm ok having this in a follow up PR. FYI @hbjORbj

I don't understand what you mean here? I've tried to abstract all hooks that were present in the shell

> Mobile nav ❌
On it! Also some type errors i need to fix around this area too

**Review by joeauyeung — APPROVED** [review]
Navigation is now working as expected. 

In terms of my abstracting hooks logic comment. We can't call hooks in server components so ideally I would have liked to see the logic inside of the hooks to be abstracted. Although looking through it some more I realize it would be too big for this PR so let's cross that bridge when we get there.

**Pull request merged**